src/rewriter/rewrite_db_proof_cons.cpp

```diff
@@ -41,8 +70,950 @@ bool RewriteDbProofCons::prove(CDProof* cdp,
+        stgt = d_currFixedPointConc[1];
+      }
+      d_currFixedPointConc = Node::null();
+    } while (continueFixedPoint);
+
+    std::vector<Node> transEq;
+    Node prev = ssrc;
+    Node context = rpr.getContext();
+    Node placeholder = context[0][0];
+    Node body = context[1];
+    Node currConc = body;
+    Node currContext = placeholder;
+    for (size_t i = 0, size = steps.size(); i < size; i++)
+    {
+      const std::vector<Node>& stepSubs = stepsSubs[i];
+      Node step = steps[i];
+      Node source = expr::narySubstitute(conc[0], vars, stepSubs);
+      Node target = expr::narySubstitute(body, vars, stepSubs);
+      target = target.substitute(TNode(placeholder), TNode(step));
+      cacheProofSubPlaceholder(currContext, placeholder, source, target);
+
+      ProvenInfo dpi;
+      dpi.d_id = pi.d_id;
+      dpi.d_vars = vars;
+      dpi.d_subs = stepSubs;
+      d_pcache[source.eqNode(target)] = dpi;
+
+      currConc = expr::narySubstitute(currConc, vars, stepSubs);
+      currContext = currConc;
+      Node prevConc = currConc;
+      if (i < size - 1)
+      {
+        currConc = currConc.substitute(TNode(placeholder), TNode(body));
+      }
+      Node stepConc = prevConc.substitute(TNode(placeholder), TNode(step));
+      transEq.push_back(prev.eqNode(stepConc));
+      prev = stepConc;
+    }
+
+    d_currFixedPointId = DslProofRule::FAIL;
+    // add the transistivity rule here if needed
+    if (transEq.size() >= 2)
+    {
+      pi.d_id = DslProofRule::TRANS;
+      // store transEq in d_vars
+      pi.d_vars = transEq;
+      // return the end of the chain, which will be used for constrained
+      // matching
+      return transEq.back()[1];
+    }
+  }
+
+  Node res = conc[1];
+  if (rpr.isFixedPoint())
+  {
+    Node context = rpr.getContext();
+    res = context[1].substitute(TNode(context[0][0]), TNode(conc[1]));
+  }
+  return expr::narySubstitute(res, vars, subs);
+}
+
+void RewriteDbProofCons::cacheProofSubPlaceholder(TNode context,
+                                                  TNode placeholder,
+                                                  TNode source,
+                                                  TNode target)
+{
+  std::vector<TNode> toVisit = {context};
+  std::unordered_map<TNode, TNode> parent;
+  std::vector<Node> congs;
+  parent[context] = TNode::null();
+  while (!toVisit.empty())
+  {
+    TNode curr = toVisit.back();
+    toVisit.pop_back();
+
+    if (curr == placeholder)
+    {
+      while (parent[curr] != Node::null())
+      {
+        Node lhs = parent[curr].substitute(placeholder, source);
+        Node rhs = parent[curr].substitute(placeholder, target);
+        congs.emplace_back(lhs.eqNode(rhs));
+        curr = parent[curr];
+      }
+      break;
+    }
+
+    for (TNode n : curr)
+    {
+      if (parent.find(n) != parent.end())
+      {
+        continue;
+      }
+
+      toVisit.emplace_back(n);
+      parent[n] = curr;
```

Comment:
```suggestion
      auto [it, inserted] = parent.emplace(n, curr);
      if (inserted)
      {
        toVisit.emplace_back(n);
      }
```

Saves again 1 map lookup.